### PR TITLE
Multiple corrections and code completion

### PR DESCRIPTION
**Pull request by oualib**
- Display Bug for autoML - corrected
 - Adding to_python for Pipelines
 - to_python is now available for NaiveBayes and OneHotEncoder
 - correction to_sql for OneHotEncoder
 - correction of read_csv for headers with space
 - correction of outlier_plot

### CODE REVIEW

File: verticapy/learn/vmodel.py [hunk 1]
```diff
@@ -3501,15 +3525,15 @@ def score(self, method: str = "accuracy", cutoff: float = 0.5):
...
+            return roc_curve(self.y, self.deploySQL(), self.test_relation, self.cursor, auc_roc=True, nbins=10000,)
         elif method in ("best_cutoff", "best_threshold"):
             return roc_curve(
                 self.y,
                 self.deploySQL(),
                 self.test_relation,
                 self.cursor,
                 best_threshold=True,
-                nbins=1000,
+                nbins=10000,
```
Comment (afard) [inline verticapy/learn/vmodel.py:3536]: Why increasing nbins? 1000 was not enough?

File: verticapy/tests/vModel/test_decision_tree_classifier.py [hunk 2]
```diff
@@ -223,6 +223,21 @@ def test_to_sklearn(self, model):
 
         # 'predict_proba'
 
+    def test_to_python(self, model, titanic_vd):
+        model_test = DecisionTreeClassifier("rfc_python_test", cursor=model.cursor)
+        model_test.drop()
+        model_test.fit(titanic_vd, ["age", "fare", "sex"], "embarked")
+        model_test.cursor.execute(
+            "SELECT PREDICT_RF_CLASSIFIER(* USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True) FROM (SELECT 30.0 AS age, 45.0 AS fare, 'male' AS sex) x"
```
Comment (afard) [inline verticapy/tests/vModel/test_decision_tree_classifier.py:231]: I think you can simplify the query to:

"SELECT PREDICT_RF_CLASSIFIER(30.0, 45.0, 'male' USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True)"

File: verticapy/tests/vModel/test_decision_tree_regressor.py [hunk 3]
```diff
@@ -231,6 +231,15 @@ def test_to_sklearn(self, base):
 
         model_sk.drop()
 
+    def test_to_python(self, model):
+        model.cursor.execute(
+            "SELECT PREDICT_RF_REGRESSOR(* USING PARAMETERS model_name = '{}', match_by_pos=True)::float FROM (SELECT 'Male' AS \"Gender\", 0 AS \"owned cars\", 'Cheap' AS \"cost\", 'Low' AS \"income\") x".format(
```
Comment (afard) [inline verticapy/tests/vModel/test_decision_tree_regressor.py:236]: You can simplify the query into:
"SELECT PREDICT_RF_REGRESSOR('Male', 0, 'Cheap', 'Low' USING PARAMETERS model_name = '{}', match_by_pos=True)::float"

File: verticapy/tests/vModel/test_dummy_tree_classifier.py [hunk 4]
```diff
@@ -213,6 +213,21 @@ def test_to_sklearn(self, model):
 
         # 'predict_proba'
 
+    def test_to_python(self, model, titanic_vd):
+        model_test = DummyTreeClassifier("rfc_python_test", cursor=model.cursor)
+        model_test.drop()
+        model_test.fit(titanic_vd, ["age", "fare", "sex"], "embarked")
+        model_test.cursor.execute(
+            "SELECT PREDICT_RF_CLASSIFIER(* USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True) FROM (SELECT 30.0 AS age, 45.0 AS fare, 'male' AS sex) x"
```
Comment (afard) [inline verticapy/tests/vModel/test_dummy_tree_classifier.py:221]: "SELECT PREDICT_RF_CLASSIFIER(30.0, 45.0, 'male' USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True)"

File: verticapy/tests/vModel/test_dummy_tree_regressor.py [hunk 5]
```diff
@@ -213,6 +213,15 @@ def test_to_sklearn(self, base):
 
         model_sk.drop()
 
+    def test_to_python(self, model):
+        model.cursor.execute(
+            "SELECT PREDICT_RF_REGRESSOR(* USING PARAMETERS model_name = '{}', match_by_pos=True)::float FROM (SELECT 'Male' AS \"Gender\", 0 AS \"owned cars\", 'Cheap' AS \"cost\", 'Low' AS \"income\") x".format(
```
Comment (afard) [inline verticapy/tests/vModel/test_dummy_tree_regressor.py:218]: "SELECT PREDICT_RF_REGRESSOR('Male', 0, 'Cheap', 'Low' USING PARAMETERS model_name = '{}', match_by_pos=True)::float"

File: verticapy/tests/vModel/test_onehotencoder.py [hunk 6]
```diff
@@ -132,6 +132,24 @@ def test_to_sql(self, model):
         prediction2 = [float(elem) for elem in model.cursor.fetchone()]
         assert prediction == pytest.approx(prediction2)
 
+    def test_to_python(self, model):
+        model.cursor.execute(
+            "SELECT pclass_0, pclass_1, pclass_2, sex_0, sex_1, embarked_0, embarked_1, embarked_2, 0 FROM (SELECT APPLY_ONE_HOT_ENCODER(pclass, sex, embarked USING PARAMETERS model_name = '{}', match_by_pos=True, drop_first=False) FROM (SELECT 1 AS pclass, 'female' AS sex, 'S' AS embarked) x) x".format(
```
Comment (afard) [inline verticapy/tests/vModel/test_onehotencoder.py:137]: "SELECT pclass_0, pclass_1, pclass_2, sex_0, sex_1, embarked_0, embarked_1, embarked_2, 0 FROM (SELECT APPLY_ONE_HOT_ENCODER(1, 'female', 'S' USING PARAMETERS model_name = '{}', match_by_pos=True, drop_first=False)"

File: verticapy/tests/vModel/test_randomforest_classifier.py [hunk 7]
```diff
@@ -223,6 +223,22 @@ def test_to_sklearn(self, model):
 
         # 'predict_proba'
 
+    def test_to_python(self, model, titanic_vd):
+        model_test = RandomForestClassifier("rfc_python_test", cursor=model.cursor)
+        model_test.drop()
+        model_test.fit(titanic_vd, ["age", "fare", "sex"], "embarked")
+        model_test.cursor.execute(
+            "SELECT PREDICT_RF_CLASSIFIER(* USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True) FROM (SELECT 30.0 AS age, 45.0 AS fare, 'male' AS sex) x"
```
Comment (afard) [inline verticapy/tests/vModel/test_randomforest_classifier.py:231]: "SELECT PREDICT_RF_CLASSIFIER(30.0, 45.0, 'male' USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True)"

File: verticapy/tests/vModel/test_randomforest_regressor.py [hunk 8]
```diff
@@ -227,6 +227,15 @@ def test_to_sklearn(self, model):
         prediction = model.cursor.fetchone()[0]
         assert prediction == pytest.approx(md.predict([["Male", 0, "Cheap", "Low"]])[0])
 
+    def test_to_python(self, model):
+        model.cursor.execute(
+            "SELECT PREDICT_RF_REGRESSOR(* USING PARAMETERS model_name = '{}', match_by_pos=True)::float FROM (SELECT 'Male' AS \"Gender\", 0 AS \"owned cars\", 'Cheap' AS \"cost\", 'Low' AS \"income\") x".format(
```
Comment (afard) [inline verticapy/tests/vModel/test_randomforest_regressor.py:232]: "SELECT PREDICT_RF_REGRESSOR('Male', 0, 'Cheap', 'Low' USING PARAMETERS model_name = '{}', match_by_pos=True)::float"

File: verticapy/tests/vModel/test_xgb_classifier.py [hunk 9]
```diff
@@ -208,6 +208,21 @@ def test_to_sklearn(self, model):
             md.predict([["Bus", "Male", 0, "Cheap", "Low"]])[0]
         )
 
+    def test_to_python(self, model, titanic_vd):
+        model_test = XGBoostClassifier("rfc_python_test", cursor=model.cursor)
+        model_test.drop()
+        model_test.fit(titanic_vd, ["age", "fare", "sex"], "embarked")
+        model_test.cursor.execute(
+            "SELECT PREDICT_XGB_CLASSIFIER(* USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True) FROM (SELECT 30.0 AS age, 45.0 AS fare, 'male' AS sex) x"
```
Comment (afard) [inline verticapy/tests/vModel/test_xgb_classifier.py:216]: "SELECT PREDICT_XGB_CLASSIFIER(30.0, 45.0, 'male' USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True)"

File: verticapy/tests/vModel/test_xgb_classifier.py [hunk 10]
```diff
@@ -208,6 +208,21 @@ def test_to_sklearn(self, model):
...
+        )
+        prediction = model_test.cursor.fetchone()[0]
+        assert prediction == model_test.to_python(return_str=False)([[30.0, 45.0, 'male']])[0]
+        model_test.cursor.execute(
+            "SELECT PREDICT_XGB_CLASSIFIER(* USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True) FROM (SELECT 30.0 AS age, 145.0 AS fare, 'female' AS sex) x"
```
Comment (afard) [inline verticapy/tests/vModel/test_xgb_classifier.py:221]: "SELECT PREDICT_XGB_CLASSIFIER(30.0, 145.0, 'female' USING PARAMETERS model_name = 'rfc_python_test', match_by_pos=True)"

File: verticapy/tests/vModel/test_xgb_regressor.py [hunk 11]
```diff
@@ -148,6 +148,15 @@ def test_drop(self, base):
         )
         assert base.cursor.fetchone() is None
 
+    def test_to_python(self, model, titanic_vd):
+        model.cursor.execute(
+            "SELECT PREDICT_XGB_REGRESSOR(* USING PARAMETERS model_name = '{}', match_by_pos=True)::float FROM (SELECT 'Male' AS \"Gender\", 0 AS \"owned cars\", 'Cheap' AS \"cost\", 'Low' AS \"income\") x".format(
```
Comment (afard) [inline verticapy/tests/vModel/test_xgb_regressor.py:153]: "SELECT PREDICT_XGB_REGRESSOR('Male', 0, 'Cheap', 'Low' USING PARAMETERS model_name = '{}', match_by_pos=True)::float"

**Comment by afard** [comment]
Hi Badr,
I have gone through the PR once and have left a few comments. Can you please address them?

File: verticapy/tests/vModel/test_decision_tree_classifier.py (hunk 2 above)
Comment (oualib) [reply verticapy/tests/vModel/test_decision_tree_classifier.py:231]: When the dataset is unbalanced, the number of bins is not enough to get a precise AUC

Comment (oualib) [reply verticapy/tests/vModel/test_decision_tree_classifier.py:231]: Done;

File: verticapy/tests/vModel/test_decision_tree_regressor.py (hunk 3 above)
Comment (oualib) [reply verticapy/tests/vModel/test_decision_tree_regressor.py:236]: Done;

File: verticapy/tests/vModel/test_dummy_tree_classifier.py (hunk 4 above)
Comment (oualib) [reply verticapy/tests/vModel/test_dummy_tree_classifier.py:221]: Done;

File: verticapy/tests/vModel/test_dummy_tree_regressor.py (hunk 5 above)
Comment (oualib) [reply verticapy/tests/vModel/test_dummy_tree_regressor.py:218]: Done;

File: verticapy/tests/vModel/test_xgb_regressor.py (hunk 11 above)
Comment (oualib) [reply verticapy/tests/vModel/test_xgb_regressor.py:153]: Done;

File: verticapy/tests/vModel/test_onehotencoder.py (hunk 6 above)
Comment (oualib) [reply verticapy/tests/vModel/test_onehotencoder.py:137]: Done;

File: verticapy/tests/vModel/test_randomforest_classifier.py (hunk 7 above)
Comment (oualib) [reply verticapy/tests/vModel/test_randomforest_classifier.py:231]: Done;

File: verticapy/tests/vModel/test_randomforest_regressor.py (hunk 8 above)
Comment (oualib) [reply verticapy/tests/vModel/test_randomforest_regressor.py:232]: Done;

File: verticapy/tests/vModel/test_xgb_classifier.py (hunk 9 above)
Comment (oualib) [reply verticapy/tests/vModel/test_xgb_classifier.py:216]: Done;

File: verticapy/tests/vModel/test_xgb_classifier.py (hunk 10 above)
Comment (oualib) [reply verticapy/tests/vModel/test_xgb_classifier.py:221]: Done;

**Comment by afard** [comment]
The TestOneHotEncoder.test_to_python test has failed.

**Review by afard — APPROVED** [review]

**Pull request merged**